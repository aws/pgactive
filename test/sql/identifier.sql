\c postgres

-- No real way to test the sysid, so ignore it
SELECT timeline = 0, dboid = (SELECT oid FROM pg_database WHERE datname = current_database())
FROM pgactive.pgactive_get_local_nodeid();

SELECT current_database() = 'postgres';

-- Test probing for replication connection and get node information of a given
-- dsn. Note that local node and remote node having different node identifiers
-- (r.sysid = l.sysid false) as each database gets unique pgactive node
-- identifier.
SELECT
	r.sysid = l.sysid,
	r.timeline = l.timeline,
	r.dboid = (SELECT oid FROM pg_database WHERE datname = 'regression'),
	variant = pgactive.pgactive_variant(),
	version = pgactive.pgactive_version(),
	version_num = pgactive.pgactive_version_num(),
	min_remote_version_num = pgactive.pgactive_min_remote_version_num(),
	has_required_privs = 't'
FROM pgactive._pgactive_get_node_info_private('dbname=regression') r,
     pgactive.pgactive_get_local_nodeid() l;

SELECT
    r.dboid = (SELECT oid FROM pg_database WHERE datname = current_database())
FROM pgactive._pgactive_get_node_info_private('dbname='||current_database()) r;

-- Verify that parsing slot names then formatting them again produces round-trip
-- output.
WITH namepairs(orig, remote_sysid, remote_timeline, remote_dboid, local_dboid, replication_name, formatted)
AS (
  SELECT
    s.slot_name, p.*, pgactive.pgactive_format_slot_name(p.remote_sysid, p.remote_timeline, p.remote_dboid, p.local_dboid, '')
  FROM pg_catalog.pg_replication_slots s,
    LATERAL pgactive.pgactive_parse_slot_name(s.slot_name) p
)
SELECT orig, formatted
FROM namepairs
WHERE orig <> formatted;

-- Check the view mapping slot names to pgactive nodes. We can't really examine the slot
-- name in the regresschecks, because it changes every run, so make sure we at least
-- find the expected nodes.
SELECT count(1) FROM (
    SELECT ns.node_name
	FROM pgactive.pgactive_nodes LEFT JOIN pgactive.pgactive_node_slots ns USING (node_name)
) q
WHERE node_name IS NULL;

-- Check to see if we can get the local node name
SELECT pgactive.pgactive_get_local_node_name() = 'node-pg';

-- Verify that creating/altering/dropping of pgactive node identifier getter
-- function is disallowed.

-- Must fail
CREATE OR REPLACE FUNCTION pgactive._pgactive_node_identifier_getter_private()
RETURNS numeric AS $$ SELECT '123456'::numeric $$
LANGUAGE SQL;

-- Must fail
ALTER FUNCTION pgactive._pgactive_node_identifier_getter_private STABLE;

-- Must fail
ALTER FUNCTION pgactive._pgactive_node_identifier_getter_private OWNER TO CURRENT_USER;

-- Must fail
ALTER FUNCTION pgactive._pgactive_node_identifier_getter_private RENAME TO alice;

-- Must fail
DROP FUNCTION pgactive._pgactive_node_identifier_getter_private();
