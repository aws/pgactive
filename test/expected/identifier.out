\c postgres
-- No real way to test the sysid, so ignore it
SELECT timeline = 0, dboid = (SELECT oid FROM pg_database WHERE datname = current_database())
FROM bdr.bdr_get_local_nodeid();
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

SELECT current_database() = 'postgres';
 ?column? 
----------
 t
(1 row)

-- Test probing for remote node information. Note that local node and remote
-- node having different node identifiers (r.sysid = l.sysid false) as each
-- database gets unique BDR node identifier.
SELECT
	r.sysid = l.sysid,
	r.timeline = l.timeline,
	r.dboid = (SELECT oid FROM pg_database WHERE datname = 'regression'),
	variant = bdr.bdr_variant(),
	version = bdr.bdr_version(),
	version_num = bdr.bdr_version_num(),
	min_remote_version_num = bdr.bdr_min_remote_version_num(),
	is_superuser = 't'
FROM bdr.bdr_get_remote_nodeinfo('dbname=regression') r,
     bdr.bdr_get_local_nodeid() l;
 ?column? | ?column? | ?column? | ?column? | ?column? | ?column? | ?column? | ?column? 
----------+----------+----------+----------+----------+----------+----------+----------
 f        | t        | t        | t        | t        | t        | t        | t
(1 row)

-- bdr.bdr_get_remote_nodeinfo can also be used to probe the local dsn
-- and make sure it works.
SELECT
    r.dboid = (SELECT oid FROM pg_database WHERE datname = current_database())
FROM bdr.bdr_get_remote_nodeinfo('dbname='||current_database()) r;
 ?column? 
----------
 t
(1 row)

-- Test probing for replication connection. Note that local node and remote
-- node having different node identifiers (r.sysid = l.sysid false) as each
-- database gets unique BDR node identifier.
SELECT
	r.sysid = l.sysid,
	r.timeline = l.timeline,
	r.dboid = (SELECT oid FROM pg_database WHERE datname = 'regression')
FROM bdr.bdr_test_replication_connection('dbname=regression') r,
     bdr.bdr_get_local_nodeid() l;
 ?column? | ?column? | ?column? 
----------+----------+----------
 f        | t        | t
(1 row)

-- Probing replication connection for the local dsn will work too
-- even though the identifier is the same.
SELECT
	r.dboid = (SELECT oid FROM pg_database WHERE datname = current_database())
FROM bdr.bdr_test_replication_connection('dbname='||current_database()) r;
 ?column? 
----------
 t
(1 row)

-- Verify that parsing slot names then formatting them again produces round-trip
-- output.
WITH namepairs(orig, remote_sysid, remote_timeline, remote_dboid, local_dboid, replication_name, formatted)
AS (
  SELECT
    s.slot_name, p.*, bdr.bdr_format_slot_name(p.remote_sysid, p.remote_timeline, p.remote_dboid, p.local_dboid, '')
  FROM pg_catalog.pg_replication_slots s,
    LATERAL bdr.bdr_parse_slot_name(s.slot_name) p
)
SELECT orig, formatted
FROM namepairs
WHERE orig <> formatted;
 orig | formatted 
------+-----------
(0 rows)

-- Check the view mapping slot names to bdr nodes. We can't really examine the slot
-- name in the regresschecks, because it changes every run, so make sure we at least
-- find the expected nodes.
SELECT count(1) FROM (
    SELECT ns.node_name
	FROM bdr.bdr_nodes LEFT JOIN bdr.bdr_node_slots ns USING (node_name)
) q
WHERE node_name IS NULL;
 count 
-------
     1
(1 row)

-- Check to see if we can get the local node name
SELECT bdr.bdr_get_local_node_name() = 'node-pg';
 ?column? 
----------
 t
(1 row)

-- Verify that creating/altering/dropping of BDR node identifier getter
-- function is disallowed.
-- Must fail
CREATE OR REPLACE FUNCTION bdr._bdr_node_identifier_getter_private()
RETURNS numeric AS $$ SELECT '123456'::numeric $$
LANGUAGE SQL;
ERROR:  creation of BDR node identifier getter function is not allowed
-- Must fail
ALTER FUNCTION bdr._bdr_node_identifier_getter_private STABLE;
ERROR:  altering of BDR node identifier getter function is not allowed
-- Must fail
ALTER FUNCTION bdr._bdr_node_identifier_getter_private OWNER TO CURRENT_USER;
ERROR:  altering of BDR node identifier getter function is not allowed
-- Must fail
ALTER FUNCTION bdr._bdr_node_identifier_getter_private RENAME TO alice;
ERROR:  altering of BDR node identifier getter function is not allowed
-- Must fail
DROP FUNCTION bdr._bdr_node_identifier_getter_private()
ERROR:  dropping of BDR node identifier getter function is not allowed
